File: src/main/java/komu/blunt/analyzer/StaticEnvironment.java
package komu.blunt.analyzer;

import komu.blunt.objects.Symbol;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static java.util.Arrays.asList;

public final class StaticEnvironment {

    private final StaticEnvironment parent;
    private final Map<Symbol, VariableInfo> variables = new HashMap<>();

    public StaticEnvironment() {
        this(null);
    }

    public StaticEnvironment(StaticEnvironment parent) {
        this.parent = parent;
    }

    public VariableReference lookup(Symbol var) {
        return lookup(var, 0);
    }

    private VariableReference lookup(Symbol name, int depth) {
        VariableInfo var = variables.get(name);
        if (var != null)
            return reference(depth, var.offset, name);
        else if (parent != null)
            return parent.lookup(name, depth+1);
        else
            throw new UnboundVariableException(name);
    }

    public VariableReference define(Symbol name) {
        if (variables.containsKey(name))
            throw new AnalyzationException("Variable " + name + " is already defined in this scope.");

        int offset = variables.size();
        variables.put(name, new VariableInfo(name, offset));
        return reference(0, offset, name);
    }
    
    public VariableReference lookupInCurrentScopeOrDefine(Symbol symbol) {
        VariableInfo var = variables.get(symbol);
        
        if (var != null) {
            return reference(0, var.offset, var.name);
        } else {
            return define(symbol);
        }
    }
    
    private VariableReference reference(int frame, int offset, Symbol name) {
        if (parent == null)
            return VariableReference.global(offset, name);
        else
            return VariableReference.nested(frame, offset, name);
    }

    public StaticEnvironment extend(Symbol symbol) {
        return extend(asList(symbol));
    }

    public StaticEnvironment extend() {
        return new StaticEnvironment(this);
    }

    public StaticEnvironment extend(List<Symbol> symbols) {
        StaticEnvironment env = new StaticEnvironment(this);

        for (Symbol symbol : symbols)
            env.define(symbol);

        return env;
    }

    public int size() {
        return variables.size();
    }
}

File: src/main/kotlin/komu/blunt/analyzer/StaticEnvironment.kt
package komu.blunt.analyzer

import komu.blunt.objects.Symbol
import java.util.*

class StaticEnvironment(private val parent: StaticEnvironment? = null) {

    private val variables = HashMap<Symbol, VariableInfo>()

    val size: Int
        get() = variables.size

    operator fun get(name: Symbol): VariableReference =
        get(name, 0)

    fun get(name: Symbol, depth: Int): VariableReference =
        variables[name]?.toReference(depth) ?:
            parent?.get(name, depth+1) ?:
            throw UnboundVariableException(name)

    fun define(name: Symbol): VariableReference {
        if (variables.containsKey(name))
            throw AnalyzationException("Variable '$name' is already defined in this scope.")

        val v = VariableInfo(name, variables.size)
        variables[name] = v
        return v.toReference(0)
    }

    fun lookupInCurrentScopeOrDefine(name: Symbol): VariableReference =
        variables[name]?.toReference(0) ?: define(name)

    private fun reference(frame: Int, offset: Int, name: Symbol): VariableReference =
        if (parent == null)
            VariableReference.global(offset, name)
        else
            VariableReference.nested(frame, offset, name)

    fun extend(vararg names: Symbol) =
        extend(names.asList())

    fun extend(symbols: Iterable<Symbol>): StaticEnvironment {
        val env = StaticEnvironment(this)

        for (symbol in symbols)
            env.define(symbol)

        return env
    }
}
